torch/distributed/checkpoint/planner_helpers.py

```diff
@@ -310,12 +307,80 @@ def _init_meta_tensor(value: Any) -> Any:
                 stride=value.stride(),
             )
             return dtensor
-        elif isinstance(value, torch.Tensor):
-            tensor = torch.empty_like(value, device=device)
-            return tensor
         else:
+            return value
+
+    def sharded_tensor_func(value: Any):
+        device = getattr(value, "device", None)
+        if device == torch.device("meta"):
             raise RuntimeError(
                 f"Found unsupported type {type(value)} for meta device loading."
             )
-    else:
-        return value
+        else:
+            return value
+
+    def tensor_func(value: torch.Tensor):
+        device = getattr(value, "device", None)
+        if device == torch.device("meta"):
+            device_type = dist.distributed_c10d._get_pg_default_device().type
+            device = cast(
+                torch.device, _get_device_module(device_type).current_device()
+            )
+            tensor = torch.empty_like(value, device=device)
+            return tensor
+        else:
+            return value
+
+    _iterate_state_dict(
+        state_dict,
+        dtensor_func,
+        sharded_tensor_func,
+        tensor_func,
+    )
+
+
+def _iterate_state_dict(
+    iter_object: Any,
+    dtensor_func: Callable,
+    sharded_tensor_func: Callable,
+    tensor_func: Callable,
+):
```

Comment:
Since our use case is narrowed to nested `dict` rather than general pytrees, I think a better implementation is to have a `deep_update` function for nested `dict`s. Then you can reuse the `tree_map` function:

```python
nested_meta_tensor_dict = ...
initialized_nested_meta_tensor_dict = pytree.tree_map(init, nested_meta_tensor_dict)
deep_update(initialized_nested_meta_tensor_dict, nested_meta_tensor_dict)
```